pkg/informant/cgroupmanager.go

```diff
@@ -0,0 +1,222 @@
+package informant
+
+// A lightweight wrapper around cgroup2.Manager, with a mix of convenience and extra functionality.
+
+import (
+	"errors"
+	"fmt"
+	"os"
+	"path/filepath"
+	"strconv"
+	"strings"
+	"sync/atomic"
+
+	cgroups "github.com/containerd/cgroups/v3"
+	"github.com/containerd/cgroups/v3/cgroup2"
+
+	"github.com/neondatabase/autoscaling/pkg/util"
+)
+
+type CgroupManager struct {
+	MemoryHighEvent util.CondChannelReceiver
+	ErrCh           <-chan error
+
+	name    string
+	manager *cgroup2.Manager
+}
+
+func NewCgroupManager(groupName string) (*CgroupManager, error) {
+	mode := cgroups.Mode()
+	if mode != cgroups.Unified && mode != cgroups.Legacy {
+		var modeString string
+		switch mode {
+		case cgroups.Unavailable:
+			modeString = "Unavailable"
+		case cgroups.Legacy:
+			modeString = "cgroups v1 ONLY"
+		default:
+			panic(fmt.Errorf("unexpected cgroups mode value %d", mode))
+		}
+
+		return nil, fmt.Errorf("cgroups v2 are not enabled, mode = %q", modeString)
+	}
+
+	manager, err := cgroup2.Load(groupName)
+	if err != nil {
+		return nil, fmt.Errorf("Error loading cgroup: %w", err)
+	}
+
+	// FIXME: There's currently no way to stop the goroutine spawned by EventChan, so it doesn't yet
+	// make sense to provide a way to cancel the goroutine to handle its events. Eventually, we
+	// should either patch containerd/cgroups or write our own implementation here.
+	memEvents, eventErrCh := manager.EventChan()
+	sendEvent, recvEvent := util.NewCondChannelPair()
+
+	highEventCount := &atomic.Uint64{}
+	errCh := make(chan error, 1)
+
+	// Long-running handler task for memory events
+	go func() {
+		for {
+			select {
```

Comment:
how does this daemon thread handle shutdown

Reply:
Refer to the comment 10 lines up:

https://github.com/neondatabase/autoscaling/blob/23e7b4c5ff02496946521d7540695e8d09b562c1/pkg/informant/cgroupmanager.go#L49-L51

And

#12:

> - [ ] Cleanup on VM informant exit
>     - note: Currently (2023-01-10), NeonVM doesn't support clean shutdown, so this is not immediately applicable

tl;dr: Currently nothing else _can_ shut down, so there isn't anything it _can_ do.